src/controller/ai_controller.c

```diff
@@ -368,56 +847,335 @@ int ai_controller_poll(controller *ctrl, ctrl_event **ev) {
+                    // persist with trying to throw
+                    a->queued_tactic = TACTIC_TRIP;
+                } else {
+                    // DEBUG("=== attempt_tactic === TACTIC_TRIP - failed");
+                    return false;
+                }
+            }
+        break;
+        default:
+            a->queued_tactic = 0;
+            return false;
+        break;
+    }
+
+    a->act_timer = 28 - (a->difficulty * 2);
+
+    controller_cmd(ctrl, a->cur_act, ev);
+    a->queued_tactic = 0;
+    return true;
+}
+
+int ai_controller_poll(controller *ctrl, ctrl_event **ev) {
+    ai *a = ctrl->data;
+    object *o = ctrl->har;
+    if (!o) {
+        return 1;
     }
+
+    har *h = object_get_userdata(o);
+
+    // Do not run AI while the game is paused
+    if(game_state_is_paused(o->gs)) { return 0; }
+
+    // Do not run AI while match is starting or ending
+    // XXX this prevents the AI from doing scrap/destruction moves
+    // XXX this could be fixed by providing a "scene changed" event
+    if(is_arena(game_state_get_scene(o->gs)->id) &&
+       arena_get_state(game_state_get_scene(o->gs)) != ARENA_STATE_FIGHTING) {
+
+        // null out selected move to fix the "AI not moving problem"
+        a->selected_move = NULL;
+        return 0;
+    }
+
+    // Grab all projectiles on screen
+    vector_clear(&a->active_projectiles);
+    game_state_get_projectiles(o->gs, &a->active_projectiles);
+
+    // Try to block har
+    if(ai_block_har(ctrl, ev)) {
+        return 0;
+    }
+
+    // Try to block projectiles
+    if(ai_block_projectile(ctrl, ev)) {
+        return 0;
+    }
+
+    // handle selected move
+    if(a->selected_move) {
+        // finish doing the selected move first
+        process_selected_move(ctrl, ev);
+        // DEBUG("=== POLL === process_selected_move");
+        return 0;
+    }
+
+    // attempt queued tactic
+    if (a->queued_tactic > 0 &&
+        (
+            h->state == STATE_STANDING ||
+            h->state == STATE_WALKTO ||
+            h->state == STATE_WALKFROM
+        ) && 
+        attempt_tactic(ctrl, ev)
+    ) {
+        // DEBUG("=== POLL === attempt_tactic passed");
+        return 0;
+    }
+
+    // attempt an attack
+    if (attempt_attack(ctrl)) {
+        // DEBUG("=== POLL === attempt_attack passed");
+        return 0;
+    }
+
+    // handle movement
+    handle_movement(ctrl, ev);
+    // DEBUG("=== POLL === handle_movement");
+
     return 0;
 }
 
-void ai_controller_create(controller *ctrl, int difficulty) {
+/** 
+ * \brief Populate pilot preferences according to perceived personality.
+ *
+ * \param pilot The pilot details.
+ * \param pilot_id An integer identifying the pilot.
+ *
+ * \return Void.
+ */
+void populate_pilot_prefs(sd_pilot *pilot, int pilot_id) {
+    // not sure if this only exists when in tournament mode
+    // but i wanted a way for arcade pilots to be behave uniquely
```

Comment:
Most likely the original game has these, but they are hardcoded to binary. This is most probably okay (at least until we later figure out the "correct" values).

Reply:
Btw, I forgot to mention -- the original game does have the pilot AI options set for tournament pilots. The values are available in our data cards here: https://katajakasa.fi/projects/openomf/cards/WORLD.TRN.html (see AI options for each tournament pilot).

Reply:
Ended up copying in the values from the tournament definition. Still hard-coded.